### ISSUE TITLE
Make parsing initializers throw errors on failure

### ISSUE THREAD

**Issue by czechboy0** [czechboy0/XcodeServerSDK#29]
Currently some of them assert, some of them fail more gracefully. This needs to be unified with the new Swift 2 error handling system.

**Comment by cojoj** [comment]
I think this one should be a more global approach and try to replace any `NSError` return pointer be replaced by `throws`. 

I suggest changing this while writing tests but we need to find a way how to test `throw`ing methods to fully cover them... My idea is to write two tests per `throw`ing method - in first test correct path and in the latter the path which `throws` exception. 

We should create new `ErrorType` `enum` which will resemble what happened and probably these methods will be optional or will return optionals.

**Comment by cojoj** [comment]
@czechboy0 I think this one should be marked as a **blocker** right now because whatever we'll try to implement we'll be basing on exisinting `XcodeServerEntity` classes.

**Comment by czechboy0** [comment]
In light of [this](https://twitter.com/jckarter/status/613783410645954560), I suggest we slow down on migrating the JSON parsing initializers to `throw`, because the code will become pretty ugly. I think we should focus on other issues before they fix it and we'll be able to do it the way it should have been all along. What do you think, @cojoj and @esttorhe?

**Comment by esttorhe** [comment]
I totally agree. The code looks cluttered and _FUGLY_ as hell; would rather keep it «failable» but clean.

**Comment by czechboy0** [comment]
This issue is now blocked, we're waiting for Swift to become sane about `throw`ing out of initializers without initializing all properties.

**Comment by esttorhe** [comment]
@czechboy0 perhaps remove the `up-for-grabs` until this is «unblocked»?

**Comment by czechboy0** [comment]
Good point :+1:

**Comment by pmkowal** [comment]
I know that we are waiting for some Swift changes now, but when I run the test below with the `"//\\https://127.0.0.1"` value as a `host` parameter I receive `EXC_BAD_ACCESS`. 

``` swift
    func testDictionaryInit() {
        XCTempAssertNoThrowError("Failed to initialize the server configuration") {
            let json = [
                "host": "//\\https://127.0.0.1",
                "user": "ICanCreateBots",
                "password": "superSecr3t"
            ]

            let config = try XcodeServerConfig(json: json)

            XCTAssertEqual(config!.host, "https://127.0.0.1", "Should create proper host address")
            XCTAssertEqual(config!.user!, "ICanCreateBots")
            XCTAssertEqual(config!.password!, "superSecr3t")
        }
    }
```

Maybe in this case we could set `JSONSerializable` init method as non failable, like below, what do you think?

``` swift
public protocol JSONSerializable {
    init(json: NSDictionary) throws
    func jsonify() -> NSDictionary
}
```

**Comment by czechboy0** [comment]
Well, we still need a way to **not** return an object if the JSON is corrupt or we don't have all required fields. Since `throw`ing is still highly impractical (because you have to initialize all variables before throwing, thus not allowing for them to be `let`s), having the initializer failable makes it possible for us to have this behavior. Hopefully Apple will fix this, so that you can `throw` at any point in initializers, so we'll be able to get rid of the `return optional` paradigm.

Btw, why does the `host` have two forward and two back slashes?

**Comment by pmkowal** [comment]
> Btw, why does the host have two forward and two back slashes?

I wanted to write some tests and I was digging through the `public required convenience init?(json: NSDictionary) throws` method of `XcodeServerConfig` class.
I noticed that when you pass the `host` parameter with two forward and two back slashes to `designated init` it throws an error, but when you pass it to `convenience init` you get `EXC_BAD_ACCESS`, which shouldn't occur IMO.

**Comment by cojoj** [comment]
@pmkowal any idea which line crashes? Have you tried setting breakpoints? I wonder which line is this bad boy... 😈

**Comment by pmkowal** [comment]
The problem is when the `public required init(var host: String, user: String?=nil, password: String?=nil) throws` method throws `ConfigurationErrors.InvalidHostProvided(host)` exception in line 84.
When I catch this exception in line 129 and return nil instead of throwing an error it behaves good (no `EXC_BAD_ACCESS`). Also when I was trying to pass the malformed `host` directly to `public required init(var host: String, user: String?=nil, password: String?=nil) throws` method, everything behaves as it should.

**Comment by cojoj** [comment]
maybe because this:

``` swift
try self.init(host: host, user: json.optionalStringForKey("user"), password: json.optionalStringForKey("password"))
```

should become this:

``` swift
do {
    try self.init(host: host, user: json.optionalStringForKey("user"), password: json.optionalStringForKey("password"))
} catch {
    Log.error("Couldn't initialize server configuration: \(error)")
}
```

❓

**Comment by cojoj** [comment]
Me and @pmkowal sat for a while on this and discussed possibilities... Seems like we're mixing _too much_... Okay, it's supported to call non-failable `init` from inside of failable `init` (_The Swift Programming Language_) - fine. But we're messing `throws` with this and I guess it's a 💥 thing... 

Here are my solutions:
- Comment out `convenince init?`
- Make `convenience init` non-failable

**Comment by czechboy0** [comment]
Is it just me or does this work fine in `struct`s, but doesn't in `class`es? Seems like we can in fact `throw` before initializing all stored properties in a `struct`.

**Comment by cojoj** [comment]
@czechboy0 it seems very likely... I will `try` it (😬) ASAP and let you know!

**Comment by cojoj** [comment]
Yeah, it work for `structs`, but I'm also _almost sure_ that this was always possible for `struct`s... ¯_(ツ)_/¯ 

![zrzut ekranu 2015-10-13 o 07 23 12](https://cloud.githubusercontent.com/assets/4522407/10446151/9220522c-717b-11e5-9637-fb6a5e8bbb0c.png)

It looks like a lot of work with converting everything to `structs` right now... Also, consider that we'll need some `protocol Entity` with `extension`s, so we can provide default implementation for some properties.
But... YEAH, I'M IN FAVOR OF DOING THIS! **Crusty would be happy!**

![](http://media4.giphy.com/media/mOKKU0WU1AxFu/giphy.gif)
